### ISSUE TITLE
Method `stop_models` gives unexpected behaviour on Docker and Kubernetes

### ISSUE THREAD

**Issue by jacekwachowiak** [ucbrise/clipper#729]
I am running the current develop branch version of Clipper.
I have encountered strange outputs while using the `clipper_conn.stop_models`.
1. On Docker the method crashes with a lengthy output but does remove the running container (the output is pasted at the bottom).

2. On Kubernetes the method returns:
```
clipper_conn.stop_models('stop-test')
19-06-10:15:43:40 INFO     [clipper_admin.py:1285] Model stop-test:1 was successfully deleted
19-06-10:15:43:40 INFO     [clipper_admin.py:1319] [default-default-cluster] Stopped all containers for these models and versions:
{'stop-test': ['1']}

clipper_conn.stop_versioned_models({"stop-test": "1"})
19-06-10:15:43:51 ERROR    [clipper_admin.py:1280] Received error status code: 400 and message: model with name 'stop-test' and version '1' does not exist

clipper_conn.get_all_models()
[]

[cloud-user@k1]$ date
Mon Jun 10 15:46:09 JST 2019
[cloud-user@k1]$ kubectl get po -o wide | grep stop
stop-test-1-deployment-at-0-at-default-cluster-7b586b67f7-7bpcb   1/1     Running   0          3m29s   10.233.96.7    node2   <none>           <none>
```
so if I understand correctly Clipper says that the model was removed correctly, but then as you can see the model pod is still running, it is orphaned and undetected by Clipper but still managed by Kubernetes.  Using `stop_all_model_containers` does nothing. Only running `stop_all` cleans the pods.

Before I could delete the Docker models without the error so I guess it must be some very recent change. I am not sure if the pod removal for Kubernetes worked though. This is the Docker error output:
```
---------------------------------------------------------------------------
RemoteDisconnected                        Traceback (most recent call last)
~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    600                                                   body=body, headers=headers,
--> 601                                                   chunked=chunked)
    602 

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    386                     # otherwise it looks like a programming error was the cause.
--> 387                     six.raise_from(e, None)
    388         except (SocketTimeout, BaseSSLError, SocketError) as e:

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/packages/six.py in raise_from(value, from_value)

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    382                 try:
--> 383                     httplib_response = conn.getresponse()
    384                 except Exception as e:

~/.pyenv/versions/3.6.8/lib/python3.6/http/client.py in getresponse(self)
   1330             try:
-> 1331                 response.begin()
   1332             except ConnectionError:

~/.pyenv/versions/3.6.8/lib/python3.6/http/client.py in begin(self)
    296         while True:
--> 297             version, status, reason = self._read_status()
    298             if status != CONTINUE:

~/.pyenv/versions/3.6.8/lib/python3.6/http/client.py in _read_status(self)
    265             # sending a valid response.
--> 266             raise RemoteDisconnected("Remote end closed connection without"
    267                                      " response")

RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

ProtocolError                             Traceback (most recent call last)
~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    448                     retries=self.max_retries,
--> 449                     timeout=timeout
    450                 )

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    638             retries = retries.increment(method, url, error=e, _pool=self,
--> 639                                         _stacktrace=sys.exc_info()[2])
    640             retries.sleep()

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/util/retry.py in increment(self, method, url, response, error, _pool, _stacktrace)
    356             if read is False or not self._is_method_retryable(method):
--> 357                 raise six.reraise(type(error), error, _stacktrace)
    358             elif read is not None:

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/packages/six.py in reraise(tp, value, tb)
    684         if value.__traceback__ is not tb:
--> 685             raise value.with_traceback(tb)
    686         raise value

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    600                                                   body=body, headers=headers,
--> 601                                                   chunked=chunked)
    602 

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    386                     # otherwise it looks like a programming error was the cause.
--> 387                     six.raise_from(e, None)
    388         except (SocketTimeout, BaseSSLError, SocketError) as e:

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/packages/six.py in raise_from(value, from_value)

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    382                 try:
--> 383                     httplib_response = conn.getresponse()
    384                 except Exception as e:

~/.pyenv/versions/3.6.8/lib/python3.6/http/client.py in getresponse(self)
   1330             try:
-> 1331                 response.begin()
   1332             except ConnectionError:

~/.pyenv/versions/3.6.8/lib/python3.6/http/client.py in begin(self)
    296         while True:
--> 297             version, status, reason = self._read_status()
    298             if status != CONTINUE:

~/.pyenv/versions/3.6.8/lib/python3.6/http/client.py in _read_status(self)
    265             # sending a valid response.
--> 266             raise RemoteDisconnected("Remote end closed connection without"
    267                                      " response")

ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
<ipython-input-22-920785050951> in <module>
      1 # Stop the model
----> 2 clipper_conn.stop_models('pytorch-mod')

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/clipper_admin/clipper_admin.py in stop_models(self, model_names)
   1313                     model_dict[m["model_name"]] = [m["model_version"]]
   1314         self.cm.stop_models(model_dict)
-> 1315         self._unregister_versioned_models(model_dict)
   1316         pp = pprint.PrettyPrinter(indent=4)
   1317         self.logger.info(

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/clipper_admin/clipper_admin.py in _unregister_versioned_models(self, model_versions_dict)
   1273                 req_json = json.dumps({"model_name": model_name,
   1274                                        "model_version": model_version})
-> 1275                 r = requests.post(url, headers=headers, data=req_json)
   1276                 logger.debug(r.text)
   1277                 if r.status_code != requests.codes.ok:

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/requests/api.py in post(url, data, json, **kwargs)
    114     """
    115 
--> 116     return request('post', url, data=data, json=json, **kwargs)
    117 
    118 

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/requests/api.py in request(method, url, **kwargs)
     58     # cases, and look like a memory leak in others.
     59     with sessions.Session() as session:
---> 60         return session.request(method=method, url=url, **kwargs)
     61 
     62 

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    531         }
    532         send_kwargs.update(settings)
--> 533         resp = self.send(prep, **send_kwargs)
    534 
    535         return resp

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/requests/sessions.py in send(self, request, **kwargs)
    644 
    645         # Send the request
--> 646         r = adapter.send(request, **kwargs)
    647 
    648         # Total elapsed time of the request (approximately)

~/.pyenv/versions/3.6.8/lib/python3.6/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    496 
    497         except (ProtocolError, socket.error) as err:
--> 498             raise ConnectionError(err, request=request)
    499 
    500         except MaxRetryError as e:

ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))
```

**Comment by withsmilo** [comment]
@jacekwachowiak 
Let me check. What version is your management_frontend image in use? Did you use old management_frontend image?

**Comment by jacekwachowiak** [comment]
Ok, I am reading now that the Docker images are not updated automatically, so I have to pull them, they are probably outdated, I will pull all of them and recheck!
UPDATE: Using the latest images works well for Docker - error fixed!
UPDATE2: I am already using the latest images with Kubernetes, the models are not stopped by Clipper.

**Comment by withsmilo** [comment]
1. On Docker the problem was resolved, right?
2. We need to debug these codes in KubernetesContainerManager,
https://github.com/ucbrise/clipper/blob/develop/clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py#L621-L638

**Comment by jacekwachowiak** [comment]
Yes, on Docker everything is fine.
I found this, but since the author did not respond we'll never know if the problem was related #490

Running `stop_models()` removes the deployment but not the pods
I have discovered a dirty trick - **setting the replicas to 0** and then running one of the `stop_models` function gives a clean result. This way the pods are not orphaned. In any other case they cannot be removed even manually and cannot be managed by the deployment since it is already deleted. 
Update: It seems that the `replicaset` intermediary is a problem.

**Comment by jacekwachowiak** [comment]
Some reference materials:
https://github.com/kubernetes/kubernetes/issues/34052
tools that had the same problem:
https://github.com/pachyderm/pachyderm/pull/1694
https://github.com/garden-io/garden/issues/379
From what I see using the API does not work in a cascade mode, deleting only the explicitly defined element, in our case the `deployment`, leaving the `replicaset` and the pods. I will look for more info if the API allows now the cascade mode, but if it does not either the `replicaset` must be removed with the function call or maybe take a look at this https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/

**Comment by jacekwachowiak** [comment]
I am currently testing something and the first result is ok, so hopefully tomorrow I can create a PR!

**Comment by withsmilo** [comment]
Good! Welcome to your PR! :laughing:

**Comment by jacekwachowiak** [comment]
I added removal of each of the replicasets but suprisingly it does not clean the pods and the https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/ExtensionsV1beta1Api.md#delete_collection_namespaced_replica_set does not allow direct pod removal so it seems my solution does not work

**Comment by withsmilo** [comment]
@jacekwachowiak 
I think we have to call [`delete_namespaced_deployment()`](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/ExtensionsV1beta1Api.md#delete_namespaced_deployment) per a query_frontend replica instead of `delete_collection_namespaced_deployment()`. Would you please test these codes?
```python
    def stop_models(self, models):
        try:
            for m in models:
                for v in models[m]:
                    for query_frontend_id in range(self.num_frontend_replicas):
                        deployment_name = get_model_deployment_name(
                            m, v, query_frontend_id, self.cluster_name)
                        self._k8s_beta.delete_namespaced_deployment(
                            namespace=self.k8s_namespace,
                            name=deployment_name,
                            body=V1DeleteOptions())
        except ApiException as e:
            self.logger.warning(
                "Exception deleting kubernetes deployments: {}".format(e))
            raise e
```

**Comment by jacekwachowiak** [comment]
Hmm, I have taken another approach. I can try with this and the `propagation_policy`, since the option we are using now does not have it and it is surely part of a problem.
I created a PR but if I find this option more concise and working the same I'll edit it asap, thanks for the link!

**Comment by withsmilo** [comment]
@jacekwachowiak 
I found a bug. We have to use `labels`, not `label`.
https://github.com/ucbrise/clipper/blob/develop/clipper_admin/clipper_admin/kubernetes/model-container-template.yaml#L4
Could this patch solve your problem?

**Comment by jacekwachowiak** [comment]
> I found a bug. We have to use `labels`, not `label`.
> https://github.com/ucbrise/clipper/blob/develop/clipper_admin/clipper_admin/kubernetes/model-container-template.yaml#L4
> Could this patch solve your problem?

No, it doesn't seem so, I haven't observed any changes when I changed only this line and left the stop functions untouched

**Comment by jacekwachowiak** [comment]
```   
def stop_models(self, models):
        # Stops all deployments of pods running Clipper models with the specified
        # names and versions.
        try:
            for m in models:
                for v in models[m]:
                    for query_frontend_id in range(self.num_frontend_replicas):
                        deployment_name = get_model_deployment_name(
                            m, v, query_frontend_id, self.cluster_name)
                        self._k8s_beta.delete_namespaced_deployment(
                            namespace=self.k8s_namespace,
                            name=deployment_name,
                            propagation_policy='Foreground')
```
something like this works. Any idea how to apply it to `stop_all_model_containers`? We need a list of models beforehand and the `get_all_models` is unavailable, it's part of the clipper_admin not of the K8s manager

**Comment by withsmilo** [comment]
@jacekwachowiak 
You're right. I didn't consider `stop_all_model_containers()`. I will review your PR soon.

**Comment by withsmilo** [comment]
resolved by #732
